chain/network/src/ibf.rs

```diff
@@ -0,0 +1,185 @@
+        hasher.write_u64(seed);
+        let new_capacity = (1 << k) + 2;
+        Self {
+            capacity: new_capacity,
+            data: vec![IbfElem::default(); new_capacity],
+            hasher,
+            k,
+            seed,
+        }
+    }
+
+    pub fn add(&mut self, elem: u64) {
+        self.adjust(elem)
+    }
+
+    pub fn remove(&mut self, elem: u64) {
+        self.adjust(elem)
+    }
+
+    fn compute_hash(&self, elem: u64) -> u64 {
+        let mut h = self.hasher.clone();
+        h.write_u64(elem);
+        h.finish()
+    }
+
+    fn adjust(&mut self, elem: u64) {
+        self.adjust_value(elem);
+    }
+
+    pub fn merge(&mut self, rhs_data: &Vec<IbfElem>, rhs_seed: u64) {
+        assert_eq!(self.data.len(), rhs_data.len());
+        assert_eq!(self.seed, rhs_seed);
+        for i in 0..self.capacity {
+            self.data[i].merge(&rhs_data[i]);
+        }
+    }
+
+    #[cfg(test)]
+    fn recover(&mut self) -> Result<Vec<u64>, &'static str> {
+        let (result, difference) = self.try_recover();
+
+        if difference != 0 {
+            for i in 0..self.capacity {
+                if self.data[i].xor_elem != 0 {
+                    println!(
+                        "{} {:?} {}",
+                        i,
+                        self.data[i],
+                        self.compute_hash(self.data[i].xor_elem)
+                    );
+                }
+            }
+            return Err("unable to recover result");
+        }
+        return Ok(result);
+    }
+
+    pub fn try_recover(&mut self) -> (Vec<u64>, u64) {
+        let mut result = Vec::with_capacity(self.capacity);
+        let mut to_check = Vec::default();
+        for i in 0..self.capacity {
+            to_check.push(i);
+
+            while let Some(i) = to_check.pop() {
+                let elem = self.data[i].xor_elem;
+                if elem == 0 && self.data[i].xor_hash == 0 {
+                    continue;
+                }
+                let elem_hash = self.compute_hash(elem);
+                if elem_hash != self.data[i].xor_hash {
+                    continue;
+                }
+
+                result.push(elem);
+                self.adjust_value2(elem, elem_hash, &mut to_check);
+            }
+        }
+        let mut elems_that_differ = 0;
+        for i in 0..self.capacity {
+            if self.data[i].xor_elem != 0 {
+                elems_that_differ += 1;
+            }
+        }
+        (result, elems_that_differ)
+    }
+
+    fn generate_idx(&mut self, elem_hash: u64) -> [usize; 3] {
+        let pos0 = elem_hash & ((1 << self.k) - 1);
+        let mut pos1 = (elem_hash >> self.k) & ((1 << self.k) - 1);
+        let mut pos2 = (elem_hash >> 2 * self.k) & ((1 << self.k) - 1);
+        if pos1 >= pos0 {
+            pos1 += 1;
+        }
+        if pos2 >= min(pos0, pos1) {
+            pos2 += 1;
+        }
+        if pos2 >= max(pos0, pos1) {
+            pos2 += 1;
+        }
```

Comment:
@mfornet There is no issue here. Capacity is equal to `2^k + NUM_HASHES - 1 = 2^k + 2`, so we will never ran into the issue of pointing outside the buffer.

I've added 2 extra elements on purpose to make probability of all triplets equal. 
Otherwise 
`
pos0 = 0
pos1 = (1<<k)-1;
pos2 = (1<<k)-1;
`
would generate the same triple as
`
pos0 = 0
pos1 = 1;
pos2 = (1<<k)-1;
`

Reply:
I see, there is no risk of index out of bounds. Probably should add a comment about it.

> I've added 2 extra elements on purpose to make probability of all triplets equal.

FYI, The probability of all triplets is still not equal. You are sampling uniformly from a set of size `(2^k)^3` and mapping to a set of size `choose(2^k + 2, 3)`. Since the latter doesn't divide the former the probability of each triplet can't be the same. 

Anyway, the distribution is uniform "enough" for this use case, and I don't think we should degrade performance for this reason by changing the modulo.

Reply:
> I see, there is no risk of index out of bounds. Probably should add a comment about it.
> 
> > I've added 2 extra elements on purpose to make probability of all triplets equal.
> 
> FYI, The probability of all triplets is still not equal. You are sampling uniformly from a set of size `(2^k)^3` and mapping to a set of size `choose(2^k + 2, 3)`. Since the latter doesn't divide the former the probability of each triplet can't be the same.
> 
> Anyway, the distribution is uniform "enough" for this use case, and I don't think we should degrade performance for this reason by changing the modulo.
I think I should have written a more detailed explanation. This is just an approximation, that's used to avoid having to do modulo, so not all triples out of `choose(2^k + 2, 3)` will occur. Though for `k=10` I estimated, at at least `99.7%` of all triples can be generated, will is good enough approximation.

Reply:
I know. I didn't state it clearly.     
Let's say we choose 3 numbers in range `0...2^k-1`, we can generate all triples inside `choose(2^k, 3)` , but we also have duplicates. However, let's say we choose those tree numbers and sort them as `v_0, v_1, v_2`. We can generate a triples as `(v_0, v_1 + 1, v_2 + 2)`. 

So we can prove that the number of triples that can be generated will be between `choose(2^k, 3)`  and `choose(2^k + 2, 3)`. So for `k=10` we will be able to generate at least `choose(2^k, 3)` / `choose(2^k + 2, 3) * 100%` ~= `99.5%` of possible triples.